### PR TITLE
Sprint wrap-up docs for postgres phase 2

### PR DESCRIPTION
**Pull request by ccooper**

### CODE REVIEW

**Review by petemoore — APPROVED** [review]

File: sprints/20200622-postgres-phase2/sprint-review.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+## Goal
+This was a more mechanical sprint than previous sprints, but we knew that going in. Our goal was to migrate the existing entities-based database storage system that mimicked the Azure JSON blob datastore to a proper relational schema where data was stored in a well-defined column structure. This was reflected in our choice of Epics for this sprint.
+
+### Epics
+1. [Create and update tools to facilitate work with new db format](https://app.zenhub.com/workspaces/taskcluster-5ed15d37c2d9744af28567dc/issues/taskcluster/scrum/3)
+2. [All DB data is stored as relational tables with columns instead of entities](https://app.zenhub.com/workspaces/taskcluster-5ed15d37c2d9744af28567dc/issues/taskcluster/scrum/3)
+
+## Outcomes
+All told, there were 28 tables to migrate to the new format. @helfi92 and @djmitche had come up with a well-documented, 2-step migration plan that made most table migrations easy. In some cases, we tried to address table structure issues that were relics of thge Azure storage, e.g. extra queue tables that existed solely to provide another index, but this was the exception rather than the rule.
```
Comment (helfi92) [inline sprints/20200622-postgres-phase2/sprint-review.md]: thge

File: sprints/20200622-postgres-phase2/sprint-review.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+By the official end of the sprint (July 24), 26 of the 28 tables had been migrated in code, and the remaining two tables had code owners who were continuing to work on their migration. I say "in code" because, as of the wtime of writing, all of the database changes have not yet been pushed to the live clusters -- community & firefox-ci -- due to delays caused by the TCW (see below) and ongoing Firefox releases that made it preferable to *not* change the underlying infrastructure unduly.
```
Comment (helfi92) [inline sprints/20200622-postgres-phase2/sprint-review.md]: wtime

File: sprints/20200622-postgres-phase2/sprint-review.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+We knew going into this sprint that is was going to be a lot of work and we scheduled 5 weeks deliberately. If the work planned TCW (see below) had proven successful, we would have been on track to finish on schedule.
```
Comment (helfi92) [inline sprints/20200622-postgres-phase2/sprint-review.md]: is was going

File: sprints/20200622-postgres-phase2/sprint-review.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+As a result of the delay introduced by the TCW, @helfi92 opted as _Product owner_ to constrain the critical path to [Epic 2](https://app.zenhub.com/workspaces/taskcluster-5ed15d37c2d9744af28567dc/issues/taskcluster/scrum/3) to ensure all the tables were migrated to the new format.
+
+At the end of the sprint, the team moved any non-table migration issues labelled “db” from the Current Sprint Backlog to the Product Backlog. We also did an explicit scan of the New Issues and Product Backlog pipelines and labelled
```
Comment (helfi92) [inline sprints/20200622-postgres-phase2/sprint-review.md]: Incomplete sentence.

File: sprints/20200622-postgres-phase2/sprint-review.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+Of course, this left us with a major queue table unmigrated. It also left us with a bunch of uncertainty around the comparability of our dev and production environments, and also the unappealing prospect of needing to schedule yet another TCW.
+
+We convened a quick post-mortem on Monday morning to examine our options:
+1. **Re-run the table upgrade during the next scheduled TCW**: This was not ideal for a number of reasons. It meant waiting (most likely) another 3 weeks until the TCW. Because the changes to the database underpin so much else, this would have meant the Taskcluster code running on the firefox-ci cluster would stagnate for 3 weeks. If we did need to deploy an emergency fix, it would require a bunch of manual work.
+2. **Take an emergency downtime to run the table upgrade ASAP**: Because we didn't know *why* the performance had been so bad, this would have been a major gamble. To get more data ehre, we did ask the CloudOps team to restore a recent db backup to a separate enviroment so we could try re-running the migration under various scenarios to try to figure out the performance profile.
```
Comment (helfi92) [inline sprints/20200622-postgres-phase2/sprint-review.md]: ehre

File: sprints/20200622-postgres-phase2/sprint-review.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+3. **Modify the db functions to migrate the table data in the background while the cluster was still running**: Because the table was so large and the timing of TCWs can be uncertain, we had actually discussed this idea while planning the sprint, but had dismissed it due to the favorable timing of the current TCW. This was deemed to be the best option of the bunch, but would require very careful review to prevent data loss in the production cluster.
+
+@djmitche set to work on the background processing and after a week of testing, testing, and more testing, had a new version ready for deployment. It would take longer, but the production database would be avialable for the duration.
```
Comment (helfi92) [inline sprints/20200622-postgres-phase2/sprint-review.md]: avialble

File: sprints/20200622-postgres-phase2/sprint-review.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+CloudOps kicked off the update process the following Monday, and 47.5 hours later, the offending queu table had been migrated with 0 downtime (if you don't count the TCW, naturally).
```
Comment (helfi92) [inline sprints/20200622-postgres-phase2/sprint-review.md]: queu

**Review by helfi92 — APPROVED** [review]
Very well written. Great work!

**Review by djmitche — APPROVED** [review]
Nice write-up!  I think the review doc is the perfect size and structure for us to look back on in 6 months or a year to refresh our memory of this sprint.

**Review by imbstack — APPROVED** [review]

**Pull request merged**